Sources/ImageFetcher/ImageFetcher.swift

```diff
@@ -133,33 +131,48 @@ public extension ImageFetcher {
             }
         }
     }
+     */
 
     /// Loads the `ImageConfiguration`. If the result of the image configuration is cached, the result will be returned immediately. Otherwise a download operation will be kicked off.
     /// - Parameter url: The url of the image to be downloaded.
     /// - Returns: The result of the image load.
-    func load(_ url: URL) async -> ImageResult {
-        await load(ImageConfiguration(url: url))
+    func load(_ url: URL) async throws -> ResultType<Image> {
+        try await load(ImageConfiguration(url: url))
     }
 
     /// Loads the `ImageConfiguration`. If the result of the image configuration is cached, the result will be returned immediately. Otherwise a download operation will be kicked off.
     /// - Parameter imageConfiguration: The configuation of the image to be downloaded.
     /// - Returns: The result of the image load.
-    func load(_ imageConfiguration: ImageConfiguration) async -> ImageResult {
-        let imageTask = await task(imageConfiguration)
+    func load(_ imageConfiguration: ImageConfiguration) async throws -> ResultType<Image> {
+        if let imageTask = tasks[imageConfiguration.key] {
+            switch imageTask.state {
+            case .pending(let task):
+                let image = try await task.value
+                return .downloaded(image)
+            case .completed(let result):
+                return result
+            }
+        }
 
-        return await withUnsafeContinuation { (continuation: UnsafeContinuation<ImageResult, Never>) -> Void in
-            workerQueue.sync {
-                if let result = imageTask.result {
-                    continuation.resume(returning: result)
-                } else {
-                    imageTask.handler = { result in
-                        continuation.resume(returning: result)
-                    }
+        let task: Task<Image, Error> = Task(priority: imageConfiguration.priority) {
+            let request = URLRequest(url: imageConfiguration.url)
+            let data = try await networking.load(request).0
+            try Task.checkCancellation()
 
-                    tasks.insert(imageTask)
-                }
+            guard let editedImage = Image(data: data)?.edit(configuration: imageConfiguration) else {
+                throw ImageError.cannotParse
             }
+            try Task.checkCancellation()
+
+            try await cache(editedImage, key: imageConfiguration)
+            return editedImage
         }
+
+        tasks[imageConfiguration.key] = ImageFetcherTask(configuration: imageConfiguration, task: task)
+
+        let image = try await task.value
+        tasks[imageConfiguration.key] = ImageFetcherTask(configuration: imageConfiguration, result: .downloaded(image))
```

Comment:
It would be more performant to add `ImageFetcherTask.complete(_ image: Image)` that would mutate its state but since `ImageFetcher.task(_:)` allows access to `ImageFetcherTask` from outside the fetcher's state isolation layer, that might be dangerous. A possible work around would be to make that method `internal` or make  the fetcher task an actor itself.

Beyond this, I'm not totally certain `ImageFetcherTask` is preferable to something like:

```swift
public actor ImageFetcher {
    public enum FetcherTask {
        case pending(Task<Image, Error>)
        case completed(ResultType<Image>)

        public var isPending: Bool {
            if case .pending = self {
                return true
            } else {
                return false
            }
        }

        public var value: ResultType<Image> {
            get async throws {
                switch self {
                case .pending(let task):
                    let image = try await task.value
                    return .downloaded(image)
                case .completed(let result):
                    return result
                }
            }
        }
    }

    private var tasks: [String: FetcherTask] = [:]
}
```